File: app/src/main/java/com/wilinz/yuetingmusic/ui/user/UserListViewModel.java
package com.wilinz.yuetingmusic.ui.user;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.wilinz.yuetingmusic.data.model.User;
import com.wilinz.yuetingmusic.data.repository.UserRepository;
import com.wilinz.yuetingmusic.event.UserChangeEvent;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.List;
import java.util.Optional;

import io.reactivex.rxjava3.core.Observable;

public class UserListViewModel extends ViewModel {

    public UserListViewModel() {
        EventBus.getDefault().register(this);
        getAllUser().subscribe();
    }

    private final MutableLiveData<List<User>> usersLiveDate = new MutableLiveData<>();

    public LiveData<List<User>> getUsersLiveDate() {
        return usersLiveDate;
    }

    public Observable<List<User>> getAllUser() {
        return UserRepository.getInstance().getAllUser()
                .doOnNext(users -> {
                    refreshingLiveData.postValue(false);
                    usersLiveDate.postValue(users);
                });
    }

    public Observable<Optional<User>> exitLogin() {
        return UserRepository.getInstance().loginOrSignupVisitorUser();
    }

    public Observable<User> changeActive(User user, boolean isActive) {
        return UserRepository.getInstance().changeActive(user, isActive);
    }

    protected MutableLiveData<Boolean> refreshingLiveData = new MutableLiveData<>();

    public LiveData<Boolean> getRefreshingLiveData() {
        return refreshingLiveData;
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void refreshUser(UserChangeEvent event){
        getAllUser().subscribe();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        EventBus.getDefault().unregister(this);
    }
}

File: app/src/main/java/com/wilinz/yuetingmusic/ui/user/UserListViewModel.kt
package com.wilinz.yuetingmusic.ui.user

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.wilinz.yuetingmusic.data.model.User
import com.wilinz.yuetingmusic.data.repository.UserRepository
import com.wilinz.yuetingmusic.event.UserChangeEvent
import io.reactivex.rxjava3.core.Observable
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*

class UserListViewModel : ViewModel() {
    private val usersLiveDate = MutableLiveData<List<User>>()
    fun getUsersLiveDate(): LiveData<List<User>> {
        return usersLiveDate
    }

    val allUser: Observable<List<User>>
        get() = UserRepository.instance!!.allUser
            .doOnNext { users: List<User> ->
                refreshingLiveData.postValue(false)
                usersLiveDate.postValue(users)
            }

    fun exitLogin(): Observable<Optional<User>> {
        return UserRepository.instance!!.loginOrSignupVisitorUser()
    }

    fun changeActive(
        user: User,
        isActive: Boolean,
        vararg rememberPassword: Boolean
    ): Observable<User> {
        return UserRepository.instance!!.changeActive(user, isActive, *rememberPassword)
    }

    protected var refreshingLiveData = MutableLiveData<Boolean>()

    init {
        EventBus.getDefault().register(this)
        allUser.subscribe()
    }

    fun getRefreshingLiveData(): LiveData<Boolean> {
        return refreshingLiveData
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun refreshUser(event: UserChangeEvent?) {
        allUser.subscribe()
    }

    override fun onCleared() {
        super.onCleared()
        EventBus.getDefault().unregister(this)
    }
}